File: src/main/java/net/ccbluex/liquidbounce/utils/misc/FallingPlayer.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils.misc;

import net.ccbluex.liquidbounce.utils.MinecraftInstance;
import net.ccbluex.liquidbounce.utils.extensions.MathExtensionsKt;
import net.minecraft.client.entity.EntityPlayerSP;
import net.minecraft.util.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.util.Vec3;
import org.jetbrains.annotations.Nullable;

public class FallingPlayer extends MinecraftInstance {

    private double x;
    private double y;
    private double z;

    private double motionX;
    private double motionY;
    private double motionZ;

    private final float yaw;

    private float strafe;
    private float forward;

    public FallingPlayer(double x, double y, double z, double motionX, double motionY, double motionZ, float yaw, float strafe, float forward) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.motionX = motionX;
        this.motionY = motionY;
        this.motionZ = motionZ;
        this.yaw = yaw;
        this.strafe = strafe;
        this.forward = forward;
    }

    public FallingPlayer(EntityPlayerSP player) {
        this(player.posX, player.posY, player.posZ, player.motionX, player.motionY, player.motionZ, player.rotationYaw, player.moveStrafing, player.moveForward);
    }

    public FallingPlayer(EntityPlayerSP player, Boolean predict) {
        this(player.posX + player.motionX, player.posY + player.motionY, player.posZ + player.motionZ, player.motionX, player.motionY, player.motionZ, player.rotationYaw, player.moveStrafing, player.moveForward);
    }

    private void calculateForTick() {
        strafe *= 0.98F;
        forward *= 0.98F;

        float v = strafe * strafe + forward * forward;

        if (v >= 0.0001f) {
            v = (float) Math.sqrt(v);

            if (v < 1f) {
                v = 1f;
            }

            v = mc.thePlayer.jumpMovementFactor / v;
            strafe = strafe * v;
            forward = forward * v;
            float f1 = (float) Math.sin(MathExtensionsKt.toRadians(yaw));
            float f2 = (float) Math.cos(MathExtensionsKt.toRadians(yaw));
            motionX += strafe * f2 - forward * f1;
            motionZ += forward * f2 + strafe * f1;
        }


        motionY -= 0.08;

        motionX *= 0.91;
        motionY *= 0.9800000190734863;
        motionY *= 0.91;
        motionZ *= 0.91;

        x += motionX;
        y += motionY;
        z += motionZ;
    }

    public CollisionResult findCollision(int ticks) {
        for (int i = 0; i < ticks; i++) {
            Vec3 start = new Vec3(x, y, z);

            calculateForTick();

            Vec3 end = new Vec3(x, y, z);

            BlockPos raytracedBlock;

            float w = mc.thePlayer.width / 2F;

            if ((raytracedBlock = rayTrace(start, end)) != null) return new CollisionResult(raytracedBlock, i);

            if ((raytracedBlock = rayTrace(start.addVector(w, 0, w), end)) != null)
                return new CollisionResult(raytracedBlock, i);
            if ((raytracedBlock = rayTrace(start.addVector(-w, 0, w), end)) != null)
                return new CollisionResult(raytracedBlock, i);
            if ((raytracedBlock = rayTrace(start.addVector(w, 0, -w), end)) != null)
                return new CollisionResult(raytracedBlock, i);
            if ((raytracedBlock = rayTrace(start.addVector(-w, 0, -w), end)) != null)
                return new CollisionResult(raytracedBlock, i);

            if ((raytracedBlock = rayTrace(start.addVector(w, 0, w / 2f), end)) != null)
                return new CollisionResult(raytracedBlock, i);
            if ((raytracedBlock = rayTrace(start.addVector(-w, 0, w / 2f), end)) != null)
                return new CollisionResult(raytracedBlock, i);
            if ((raytracedBlock = rayTrace(start.addVector(w / 2f, 0, w), end)) != null)
                return new CollisionResult(raytracedBlock, i);
            if ((raytracedBlock = rayTrace(start.addVector(w / 2f, 0, -w), end)) != null)
                return new CollisionResult(raytracedBlock, i);

        }

        return null;
    }


    @Nullable
    private BlockPos rayTrace(Vec3 start, Vec3 end) {
        MovingObjectPosition result = mc.theWorld.rayTraceBlocks(start, end, true);

        if (result != null && result.typeOfHit == MovingObjectPosition.MovingObjectType.BLOCK && result.sideHit == EnumFacing.UP) {
            return result.getBlockPos();
        }

        return null;
    }

    public static class CollisionResult {
        private final BlockPos pos;
        private final int tick;

        public CollisionResult(BlockPos pos, int tick) {
            this.pos = pos;
            this.tick = tick;
        }

        public BlockPos getPos() {
            return pos;
        }

        public int getTick() {
            return tick;
        }
    }

}

File: src/main/java/net/ccbluex/liquidbounce/utils/misc/FallingPlayer.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils.misc

import net.ccbluex.liquidbounce.utils.MinecraftInstance
import net.ccbluex.liquidbounce.utils.extensions.plus
import net.ccbluex.liquidbounce.utils.extensions.toRadians
import net.minecraft.client.entity.ClientPlayerEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.Direction
import net.minecraft.util.hit.BlockHitResult.Type.BLOCK
import net.minecraft.util.math.Vec3d
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class FallingPlayer(
    private var x: Double = mc.player.x,
    private var y: Double = mc.player.y,
    private var z: Double = mc.player.z,
    private var velocityX: Double = mc.player.velocityX,
    private var velocityY: Double = mc.player.velocityY,
    private var velocityZ: Double = mc.player.velocityZ,
    private val yaw: Float = mc.player.yaw,
    private var strafe: Float = mc.player.moveStrafing,
    private var forward: Float = mc.player.moveForward
) : MinecraftInstance() {
    constructor(player: PlayerEntity, predict: Boolean = false) : this(
        if (predict) player.x + player.velocityX else player.x,
        if (predict) player.y + player.velocityY else player.y,
        if (predict) player.z + player.velocityZ else player.z,
        player.velocityX,
        player.velocityY,
        player.velocityZ,
        player.yaw,
        player.moveStrafing,
        player.moveForward
    )

    private fun calculateForTick() {
        strafe *= 0.98f
        forward *= 0.98f

        var v = strafe * strafe + forward * forward
        if (v >= 0.0001f) {
            v = mc.player.jumpMovementFactor / sqrt(v).coerceAtLeast(1f)

            strafe *= v
            forward *= v

            val f1 = sin(yaw.toRadians())
            val f2 = cos(yaw.toRadians())

            velocityX += (strafe * f2 - forward * f1).toDouble()
            velocityZ += (forward * f2 + strafe * f1).toDouble()
        }

        velocityY -= 0.08
        velocityX *= 0.91
        velocityY *= 0.9800000190734863
        velocityY *= 0.91
        velocityZ *= 0.91

        x += velocityX
        y += velocityY
        z += velocityZ
    }

    fun findCollision(ticks: Int): CollisionResult? {
        repeat(ticks) { i ->
            val start = Vec3d(x, y, z)
            calculateForTick()
            val end = Vec3d(x, y, z)

            for (offset in offsets) {
                rayTrace(start + offset, end)?.let { return CollisionResult(it, i) }
            }
        }
        return null
    }

    private fun rayTrace(start: Vec3d, end: Vec3d): BlockPos? {
        val result = mc.world.rayTrace(start, end, true) ?: return null

        return if (result.typeOfHit == BLOCK && result.sideHit == Direction.UP) result.blockPos
        else null
    }

    private val offsets = listOf(
        Vec3d(0.0, 0.0, 0.0),
        Vec3d(0.3, 0.0, 0.3),
        Vec3d(-0.3, 0.0, 0.3),
        Vec3d(0.3, 0.0, -0.3),
        Vec3d(-0.3, 0.0, -0.3),
        Vec3d(0.3, 0.0, 0.15),
        Vec3d(-0.3, 0.0, 0.15),
        Vec3d(0.15, 0.0, 0.3),
        Vec3d(0.15, 0.0, -0.3)
    )

    class CollisionResult(val pos: BlockPos, val tick: Int)
}